src/http.cc

```diff
@@ -1589,29 +1596,45 @@ HttpStateData::mayReadVirginReplyBody() const
 void
 HttpStateData::maybeReadVirginBody()
 {
-    // too late to read
-    if (!Comm::IsConnOpen(serverConnection) || fd_table[serverConnection->fd].closing())
+    if (!Comm::IsConnOpen(serverConnection) || fd_table[serverConnection->fd].closing()) {
+        debugs(11, 3, "no, peer connection gone");
         return;
+    }
+
+    if (eof) {
+        // tolerate hypothetical calls between Comm::ENDFILE and closeServer()
+        debugs(11, 3, "no, saw EOF");
+        return;
+    }
+
+    if (lastChunk) {
+        // tolerate hypothetical calls between setting lastChunk and clearing serverConnection
+        debugs(11, 3, "no, saw last-chunk");
+        return;
+    }
 
     if (!canBufferMoreReplyBytes()) {
         abortTransaction("more response bytes required, but the read buffer is full and cannot be drained");
         return;
     }
 
-    // XXX: get rid of the do_next_read flag
-    // check for the proper reasons preventing read(2)
-    if (!flags.do_next_read)
+    if (waitingForDelayAwareReadChance) {
+        debugs(11, 5, "no, still waiting for noteDelayAwareReadChance()");
         return;
+    }
 
-    flags.do_next_read = false;
+    if (waitingForCommRead) {
+        debugs(11, 3, "no, already waiting for readReply()");
+        return;
+    }
```

Comment:
This PR removes the flags.do_next_read hack from HttpStateData. Old maybeReadVirginBody() method did not Comm::Read() if that flag was false. Now that flag check is replaced with a few new conditions (seen in the diff above). AFAICT, one can pretty much ignore "flags.do_next_read = true" lines removed by this PR (they were obviously not _preventing_ any reads before and are gone now) but make sure that every removed "flags.do_next_read = false" line has the corresponding (new) condition that would prevent Squid from reaching Comm::Read().

For example, if we cleared do_next_read when encountering EOF, then the above conditions should check whether EOF was encountered and quit before the Comm::Read() call below.